tests/test-class-wc-payment-gateway-wcpay-subscriptions-process-payment.php

```diff
@@ -0,0 +1,359 @@
+			->willReturn( $this->payment_intent );
+
+		$this->mock_token_service
+			->expects( $this->once() )
+			->method( 'add_payment_method_to_user' )
+			->with( self::PAYMENT_METHOD_ID, $order->get_user() )
+			->willReturn( $this->token );
+
+		// Expect add token to order to be called, so it can be reused in renewals.
+		$this->mock_wcpay_gateway
+			->expects( $this->once() )
+			->method( 'add_token_to_order' )
+			->with(
+				$this->callback( $this->match_order_id( $order->get_id() ) ),
+				$this->token
+			);
+
+		$result       = $this->mock_wcpay_gateway->process_payment( $order->get_id() );
+		$result_order = wc_get_order( $order->get_id() );
+
+		$this->assertEquals( 'processing', $result_order->get_status() );
+		$this->assertEquals( 'success', $result['result'] );
+	}
+
+	public function test_new_card_zero_dollar_subscription() {
+		$order = WC_Helper_Order::create_order( self::USER_ID, 0 );
+
+		$this->mock_wcs_order_contains_subscription( true );
+
+		$this->mock_api_client
+			->expects( $this->once() )
+			->method( 'create_setup_intent' )
+			->with( self::PAYMENT_METHOD_ID, self::CUSTOMER_ID, 'true' )
+			->willReturn( $this->setup_intent );
+
+		$this->mock_token_service
+			->expects( $this->once() )
+			->method( 'add_payment_method_to_user' )
+			->with( self::PAYMENT_METHOD_ID, $order->get_user() )
+			->willReturn( $this->token );
+
+		// Expect add token to order to be called, so it can be reused in renewals.
+		$this->mock_wcpay_gateway
+			->expects( $this->once() )
+			->method( 'add_token_to_order' )
+			->with(
+				$this->callback( $this->match_order_id( $order->get_id() ) ),
+				$this->token
+			);
+
+		$result       = $this->mock_wcpay_gateway->process_payment( $order->get_id() );
+		$result_order = wc_get_order( $order->get_id() );
+
+		$this->assertEquals( 'processing', $result_order->get_status() );
+		$this->assertEquals( 'success', $result['result'] );
+	}
+
+	public function test_new_card_is_added_before_status_update() {
+		$order = WC_Helper_Order::create_order( self::USER_ID, 0 );
+
+		$this->mock_wcs_order_contains_subscription( true );
+
+		$this->mock_api_client
+			->expects( $this->once() )
+			->method( 'create_setup_intent' )
+			->with( self::PAYMENT_METHOD_ID, self::CUSTOMER_ID, 'true' )
+			->willReturn( $this->setup_intent );
+
+		$this->mock_token_service
+			->expects( $this->once() )
+			->method( 'add_payment_method_to_user' )
+			->with( self::PAYMENT_METHOD_ID, $order->get_user() )
+			->willReturn( $this->token );
+
+		// Expect add token to order to be called, so it can be reused in renewals.
+		$this->mock_wcpay_gateway
+			->expects( $this->once() )
+			->method( 'add_token_to_order' )
+			->with(
+				$this->logicalAnd(
+					$this->callback( $this->match_order_id( $order->get_id() ) ),
+					$this->callback( $this->match_order_status( 'pending' ) )
+				),
+				$this->token
+			);
+
+		$result = $this->mock_wcpay_gateway->process_payment( $order->get_id() );
+	}
```

Comment:
Aren't we missing some assertions here?

Reply:
The main assertion here is this one, which expects the `add_token_to_order` method to be called once, for order with ID equals to `$order->get_id()` while it's still in the `pending` status. I didn't make any assertions on the final order status because they're not relevant to test whether the card was added before the order status update, but we could include them if you'd like.

```
// Expect add token to order to be called, so it can be reused in renewals.
$this->mock_wcpay_gateway
	->expects( $this->once() )
	->method( 'add_token_to_order' )
	->with(
		$this->logicalAnd(
			$this->callback( $this->match_order_id( $order->get_id() ) ),
			$this->callback( $this->match_order_status( 'pending' ) )
		),
		$this->token
	);
```

Reply:
Ah, I totally missed the `expects` call. Looking good then, thanks!